e2e-correctness-tests/test_ursa_query.py

```diff
@@ -0,0 +1,66 @@
+import unittest
+import zmq  # type: ignore
+import json
+import os
+import logging
+
+current_path = os.path.abspath(os.path.dirname(__file__))
+testdir = current_path + "/yararules/"
+filedir = current_path + "/testfiles/"
+
+
+class TestUrsaQuery(unittest.TestCase):
+    def test_ursa_query(self):
+
+        ursa_query_files = [f for f in os.listdir(testdir) if ".txt" in f]
+
+        assert ursa_query_files
+
+        context = zmq.Context()
+        socket = context.socket(zmq.REQ)
+        logging.info("Connecting to UrsaDB...")
+        socket.connect("tcp://0.0.0.0:9281")
+        logging.info("Connected...")
+        socket.send_string('index "' + filedir + '" ' "with [gram3, text4, hash4, wide8];")
+        assert json.loads(socket.recv_string()).get("result").get("status") == "ok"
+
+        failures = []
+        for query_file in ursa_query_files:
+            logging.info("Query for: " + str(query_file))
+            with open(testdir + query_file) as f:
+                data = f.read()
+
+            logging.info("Query: " + str(data))
+            socket.send_string(data)
+
+            resp = json.loads(socket.recv_string()).get("result").get("files")
+            logging.info("Files: " + str(resp))
+
+            if (query_file[:-4] == "negative") and (len(resp) != 0):
+                logging.info(
+                    "Test failed for "
+                    + str(query_file)
+                    + ". "
+                    + str(len(resp))
+                    + " files found: "
+                    + str(resp)
+                )
+                failures.append(query_file)
+            elif (query_file[:-4] != "negative") and (
+                (len(resp) != 1) or (query_file[:-4] not in resp[0])
```

Comment:
why `!= 1`? multiple results are OK I think, but maybe I don't understand this check.

Reply:
I assumed that there will be only one file for corresponding yara rule. Then the test is simple. Check if there is only one corresponding file found. There could be also similar file that will be wrongly classified as matching the rule. What then? I could also store somewhere the number of files that should be found? Another issue is that the number can be correct but completely different wrongly matched files could be found. Maybe the probability of this is low but still exists... It's up to you what could be done...

Reply:
Another idea is to map beforehand (e.g. using python dictionary) yara rules to files that match the rule. In order to achieve this, I should know what those files are...